pkg/virtctl/create/vm.go

```diff
@@ -0,0 +1,321 @@
+			APIVersion: v1.VirtualMachineGroupVersionKind.GroupVersion().String(),
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: c.name,
+		},
+		Spec: v1.VirtualMachineSpec{
+			RunStrategy: &runStrategy,
+			Template: &v1.VirtualMachineInstanceTemplateSpec{
+				Spec: v1.VirtualMachineInstanceSpec{
+					TerminationGracePeriodSeconds: &c.terminationGracePeriod,
+				},
+			},
+		},
+	}
+}
+
+func splitKindName(kindName string) (string, string, error) {
+	split := strings.Split(kindName, "/")
+	switch l := len(split); l {
+	case 1:
+		return "", split[0], nil
+	case 2:
+		return strings.ToLower(split[0]), split[1], nil
+	default:
+		return "", "", fmt.Errorf("Invalid count %d of slashes in kind/name", l)
+	}
+}
+
+func withInstancetype(c *createVM, vm *v1.VirtualMachine) error {
+	kind, name, err := splitKindName(c.instancetype)
+	if err != nil {
+		return err
+	}
+
+	if kind != "" && kind != instancetype.SingularResourceName && kind != instancetype.ClusterSingularResourceName {
+		return fmt.Errorf("Invalid instancetype kind %s", kind)
+	}
+
+	vm.Spec.Instancetype = &v1.InstancetypeMatcher{
+		Name: name,
+		Kind: kind,
+	}
+
+	return nil
+}
+
+func withPreference(c *createVM, vm *v1.VirtualMachine) error {
+	kind, name, err := splitKindName(c.preference)
+	if err != nil {
+		return err
+	}
+
+	if kind != "" && kind != instancetype.SingularPreferenceResourceName && kind != instancetype.ClusterSingularPreferenceResourceName {
+		return fmt.Errorf("Invalid preference kind %s", kind)
+	}
+
+	vm.Spec.Preference = &v1.PreferenceMatcher{
+		Name: name,
+		Kind: kind,
+	}
+
+	return nil
+}
+
+func withImageSource(c *createVM, vm *v1.VirtualMachine) error {
+	kind, name, namespace, err := parseImageSource(c.imageSource)
+	if err != nil {
+		return err
+	}
+
+	vm.Spec.DataVolumeTemplates = []v1.DataVolumeTemplateSpec{
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: name,
+			},
+			Spec: cdiv1.DataVolumeSpec{
+				Storage: &cdiv1.StorageSpec{},
+				SourceRef: &cdiv1.DataVolumeSourceRef{
+					Kind: kind,
+					Name: name,
+				},
+			},
+		},
+	}
+	if namespace != "" {
+		vm.Spec.DataVolumeTemplates[0].Spec.SourceRef.Namespace = &namespace
+	}
+
+	vm.Spec.Template.Spec.Volumes = []v1.Volume{
+		{
+			Name: name,
+			VolumeSource: v1.VolumeSource{
+				DataVolume: &v1.DataVolumeSource{
+					Name: name,
+				},
+			},
+		},
+	}
+
+	return nil
+}
+
+func parseImageSource(imageSource string) (kind, name, namespace string, err error) {
+	split := strings.Split(imageSource, "/")
+	if l := len(split); l != 2 {
+		return "", "", "", fmt.Errorf("Invalid count %d of slashes in image source arguments", l)
+	}
+
+	switch k := strings.ToLower(split[0]); k {
+	case "pvc", "persistentvolumeclaim":
+		kind = "PersistentVolumeClaim"
+	case "dv", "datavolume":
+		kind = "DataVolume"
+	case "ds", "datasource":
+		kind = "DataSource"
+	default:
+		return "", "", "", fmt.Errorf("Invalid kind %s for image source", k)
+	}
```

Comment:
My turn to point out that `SourceRef` only supports `DataSource` at the moment:

http://kubevirt.io/api-reference/main/definitions.html#_v1beta1_datavolumesourceref

I assume you want to use `DataVolumeSourcePVC` for `PVCs`?

I'm not sure we need support for `DataVolumes` here in the indirect path?

Reply:
Touché :) That's obviously right.